File: HuMus/src/app/services/bed.service.ts
import { Injectable } from '@angular/core';
import { HttpClient, HttpHeaders } from '@angular/common/http';
import { Observable } from 'rxjs';

import { AddBedRequest, UpdateBedRequest, Bed } from '../models/bed.requests';

@Injectable({
  providedIn: 'root',
})
export class BedService {
  private baseUrl = 'http://localhost:8200/api/beds';

  constructor(private httpClient: HttpClient) {}

  addBed(authorizationHeader: string, addBedRequest: AddBedRequest): Observable<Bed> {
    const headers = new HttpHeaders({ 'Authorization': authorizationHeader });
    return this.httpClient.post<Bed>(`${this.baseUrl}/add`, addBedRequest, { headers });
  }

  updateBed(id: string, updateBedRequest: UpdateBedRequest): Observable<Bed> {
    return this.httpClient.put<Bed>(`${this.baseUrl}/update/${id}`, updateBedRequest);
  }

  findBedById(id: string): Observable<Bed> {
    return this.httpClient.get<Bed>(`${this.baseUrl}/findById/${id}`);
  }

  getAllBeds(): Observable<Bed[]> {
    return this.httpClient.get<Bed[]>(`${this.baseUrl}/findAll`);
  }

  getNearbyBeds(authorizationHeader: string, pincode: number): Observable<Bed[]> {
    const headers = new HttpHeaders({ 'Authorization': authorizationHeader });
    return this.httpClient.get<Bed[]>(`${this.baseUrl}/findNearby/${pincode}`, { headers });
  }

  getBedsByType(bedType: string): Observable<Bed[]> {
    return this.httpClient.get<Bed[]>(`${this.baseUrl}/findByType/${bedType}`);
  }

  getBedsByHospitalId(hospitalId: string): Observable<Bed[]> {
    return this.httpClient.get<Bed[]>(`${this.baseUrl}/findByHospital/${hospitalId}`);
  }

  bookBed(id: string): Observable<Bed> {
    return this.httpClient.put<Bed>(`${this.baseUrl}/bookBed/${id}`, {});
  }

  unbookBed(id: string): Observable<Bed> {
    return this.httpClient.put<Bed>(`${this.baseUrl}/unbookBed/${id}`, {});
  }

  makeBedAvailable(id: string): Observable<Bed> {
    return this.httpClient.put<Bed>(`${this.baseUrl}/makeBedAvailable/${id}`, {});
  }
}

File: Hospital Management System/Bedms/src/main/java/com/hms/bedms/service/BedService.java
package com.hms.bedms.service;

import java.util.ArrayList;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.http.HttpStatus;
import org.springframework.stereotype.Service;

import com.hms.bedms.clients.IHospitalServiceClient;
import com.hms.bedms.dtos.AddBedRequest;
import com.hms.bedms.dtos.Hospital;
import com.hms.bedms.dtos.UpdateBedRequest;
import com.hms.bedms.entities.Bed;
import com.hms.bedms.entities.Bed.BedStatus;
import com.hms.bedms.entities.Bed.BedType;
import com.hms.bedms.exceptions.BedNotFoundException;
import com.hms.bedms.exceptions.BedStatusInvalidException;
import com.hms.bedms.exceptions.HospitalNotFoundException;
import com.hms.bedms.repository.IBedRepository;

@Service
public class BedService implements IBedService {

	@Autowired
	private IBedRepository bedRepository;

	@Autowired
	private IHospitalServiceClient hospitalServiceClient;

	@Override
	public Bed addBed(String authorizationHeader, AddBedRequest addBedRequest) throws HospitalNotFoundException {
		String hospitalId = addBedRequest.getHospitalId();

		if (!(hospitalServiceClient.isHospitalFound(authorizationHeader, hospitalId)
				.getStatusCode() == HttpStatus.OK)) {
			throw new HospitalNotFoundException(hospitalId);
		}

		Bed bed = new Bed();
		bed.setHospitalId(hospitalId);
		bed.setBedType(BedType.valueOf(addBedRequest.getBedType()));
		bed.setBedStatus(BedStatus.AVAILABLE);
		bed.setCostPerDay(addBedRequest.getCostPerDay());

		return bedRepository.save(bed);
	}

	@Override
	public Bed updateBed(String id, UpdateBedRequest updateBedRequest) throws BedNotFoundException {
		Bed bed = findBedById(id);

		bed.setBedType(BedType.valueOf(updateBedRequest.getBedType()));
		bed.setCostPerDay(updateBedRequest.getCostPerDay());

		return bedRepository.save(bed);
	}

	@Override
	public Bed findBedById(String id) throws BedNotFoundException {
		return bedRepository.findById(id).orElseThrow(() -> new BedNotFoundException(id));
	}

	@Override
	public List<Bed> getAllBeds() {
		return bedRepository.findAll();
	}

	@Override
	public List<Bed> getNearbyBeds(String authorizationHeader, Long pincode) {
		List<Bed> beds = new ArrayList<>();
		List<Hospital> nearbyHospitals = hospitalServiceClient.getNearbyHospitals(authorizationHeader, pincode);
		for (Hospital hospital : nearbyHospitals) {
			String hospitalId = hospital.getId();
			beds.addAll(getBedsByHospitalId(hospitalId));
		}
		return beds;
	}

	@Override
	public List<Bed> getBedsByType(String bedType) {
		return bedRepository.findByBedType(bedType);
	}

	@Override
	public List<Bed> getBedsByHospitalId(String hospitalId) {
		return bedRepository.findByHospitalId(hospitalId);
	}

	@Override
	public Bed bookBed(String id) throws BedNotFoundException, BedStatusInvalidException {
		Bed bed = findBedById(id);
		if (bed.getBedStatus() != BedStatus.AVAILABLE) {
			throw new BedStatusInvalidException(bed.getBedStatus().toString());
		}
		bed.setBedStatus(BedStatus.BOOKED);

		return bedRepository.save(bed);
	}

	@Override
	public Bed unbookBed(String id) throws BedNotFoundException, BedStatusInvalidException {
		Bed bed = findBedById(id);
		if (bed.getBedStatus() != BedStatus.BOOKED) {
			throw new BedStatusInvalidException(bed.getBedStatus().toString());
		}
		bed.setBedStatus(BedStatus.CANCELLED);
		return bedRepository.save(bed);
	}

	@Override
	public Bed makeBedAvailable(String id) throws BedNotFoundException, BedStatusInvalidException {
		Bed bed = findBedById(id);
		if (bed.getBedStatus() != BedStatus.CANCELLED) {
			throw new BedStatusInvalidException(bed.getBedStatus().toString());
		}
		bed.setBedStatus(BedStatus.AVAILABLE);
		return bedRepository.save(bed);
	}

	public Bed completeBed(String id) throws BedNotFoundException, BedStatusInvalidException {
		Bed bed = findBedById(id);
		if (bed.getBedStatus() != BedStatus.BOOKED) {
			throw new BedStatusInvalidException(bed.getBedStatus().toString());
		}
		bed.setBedStatus(BedStatus.COMPLETED);
		return bedRepository.save(bed);
	}
}
